pjmedia/src/pjmedia-codec/lyra.cpp

```diff
@@ -0,0 +1,715 @@
+        PJ_LOG(2, (THIS_FILE, "Could not create lyra encoder"));
+        pj_mutex_unlock(lyra_data->mutex);
+        return PJMEDIA_CODEC_EFAILED;
+    }
+
+    lyra_data->dec = LyraDecoder::Create(attr->info.clock_rate,
+                                         attr->info.channel_cnt,
+                                         model_path);
+    if (lyra_data->dec == nullptr) {
+        PJ_LOG(2, (THIS_FILE, "Could not create lyra decoder"));
+        pj_mutex_unlock(lyra_data->mutex);
+        return PJMEDIA_CODEC_EFAILED;
+    }
+    lyra_data->samples_per_frame =
+                          attr->info.clock_rate / lyra_data->enc->frame_rate();
+
+    PJ_LOG(4, (THIS_FILE, "Codec opened, model_path=%.*s, chan_cnt=%d, "
+            "enc_bit_rate=%d, dec_bit_rate=%d, clockrate=%d, vad=%d, "
+            "frame_rate=%d, samples_per_frame=%d",
+            (int)lyra_cfg.model_path.slen, lyra_cfg.model_path.ptr,
+            attr->info.channel_cnt, lyra_data->enc_bit_rate,
+            lyra_data->dec_bit_rate, attr->info.clock_rate,
+            lyra_data->vad_enabled, lyra_data->enc->frame_rate(),
+            lyra_data->samples_per_frame));
+
+    pj_mutex_unlock(lyra_data->mutex);
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_close(pjmedia_codec *codec)
+{
+    struct lyra_data *lyra_data = (struct lyra_data*)codec->codec_data;
+
+    /* Destroy encoder*/
+    if (lyra_data->enc) {
+        lyra_data->enc.reset();
+    }
+
+    /* Destroy decoder */
+    if (lyra_data->dec) {
+        lyra_data->dec.reset();
+    }
+
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_modify(pjmedia_codec *codec,
+                                     const pjmedia_codec_param *attr)
+{
+    struct lyra_data *lyra_data = (struct lyra_data*)codec->codec_data;
+
+    lyra_data->plc_enabled = (attr->setting.plc != 0);
+    lyra_data->vad_enabled = (attr->setting.vad != 0);
+
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_parse(pjmedia_codec *codec,
+                                    void *pkt,
+                                    pj_size_t pkt_size,
+                                    const pj_timestamp *ts,
+                                    unsigned *frame_cnt,
+                                    pjmedia_frame frames[])
+{
+    unsigned count = 0;
+    struct lyra_data* lyra_data = (struct lyra_data*)codec->codec_data;
+    unsigned frm_size = lyra_data->dec_bit_rate / (50 * 8);
+
+    PJ_UNUSED_ARG(codec);
+
+    PJ_ASSERT_RETURN(ts && frame_cnt && frames, PJ_EINVAL);
+
+    pj_mutex_lock(lyra_data->mutex);
+    while (pkt_size >= frm_size && count < *frame_cnt) {
+        frames[count].type = PJMEDIA_FRAME_TYPE_AUDIO;
+        frames[count].buf = pkt;
+        frames[count].size = frm_size;
+        frames[count].timestamp.u64 = ts->u64 +
+                                          lyra_data->samples_per_frame * count;
+        pkt = ((char*)pkt) + frm_size;
+        pkt_size -= frm_size;
+
+        ++count;
+    }

```

Comment:
Perhaps it is better to be a bit more lenient when remote does not follow bitrate param in SDP, e.g: when `count==1` and `frm_size < pkt_size`, set the `frames[0].size = pkt_size`?

Reply:
Sure, the idea is to avoid truncating the payload if it contains only one frame.